algebra-kernels/src/polycommit.rs

```diff
@@ -0,0 +1,183 @@
+use algebra_cl_gen::gpu::{GPUError, GPUResult, kernel_polycommit, get_prefix_map};
+use algebra::{
+    AffineCurve, PrimeField
+};
+use log::{error};
+use rust_gpu_tools::*;
+
+use lazy_mut::LazyMut;
+use std::any::TypeId;
+use std::collections::HashMap;
+use std::env;
+use std::cmp;
+
+pub fn get_gpu_min_length() -> usize {
+
+    env::var("POLYCOMMIT_GPU_MIN_LENGTH")
+        .and_then(|v| match v.parse() {
+            Ok(val) => Ok(val),
+            Err(_) => {
+                error!("Invalid POLYCOMMIT_GPU_MIN_LENGTH! Defaulting to 1024...");
+                Ok(1024)
+            }
+        })
+        .unwrap_or(1024)
+}
+
+static mut CACHED_PROGRAMS: LazyMut<HashMap<opencl::Device, HashMap<TypeId, opencl::Program>>> = LazyMut::Init(HashMap::<opencl::Device, HashMap<TypeId, opencl::Program>>::new);
+
+pub struct SinglePolycommitKernel<G>
```

Comment:
Maybe we can avoid some code duplication by creating a Kernel<G> trait with functions:

fn create(d: opencl::Device) -> GPUResult<Self>;
fn get_kernels() -> GPUResult<Vec<Self>>;

And grouping some code in other functions, like the unsafe block (if needed) in the create() function, the CACHED_PROGRAMS map and so on

Reply:
More general trait with caching feature is added